In [22]:
### workclass occupation native country
# ? , 삭제, 최빈값
import pandas as pd
import numpy as np

In [23]:
train = pd.read_csv('data1/train.csv',index_col=[0])
test = pd.read_csv('data1/test.csv',index_col=[0])

In [24]:
train

age          workclass  fnlwgt      education  education-num  \
no                                                                    
1       25            Private  219199           11th              7   
2       39            Private   52978   Some-college             10   
3       35            Private  196899      Bachelors             13   
4       64            Private  135527      Assoc-voc             11   
5       24            Private   60783   Some-college             10   
...    ...                ...     ...            ...            ...   
29301   20            Private  100605        HS-grad              9   
29302   21            Private  372636        HS-grad              9   
29303   18   Self-emp-not-inc  258474           10th              6   
29304   33            Private  157446           11th              7   
29305   65                  ?   94809        HS-grad              9   

            marital-status          occupation    relationship  \
no                                                               
1                 Divorced   Machine-op-inspct   Not-in-family   
2                 Divorced       Other-service   Not-in-family   
3            Never-married   Handlers-cleaners   Not-in-family   
4                 Divorced        Tech-support   Not-in-family   
5       Married-civ-spouse    Transport-moving         Husband   
...                    ...                 ...             ...   
29301        Never-married               Sales       Own-child   
29302        Never-married               Sales       Own-child   
29303        Never-married     Farming-fishing       Own-child   
29304        Never-married        Craft-repair   Not-in-family   
29305              Widowed                   ?   Not-in-family   

                      race      sex  capital-gain  capital-loss  \
no                                                                
1                    White     Male             0             0   
2                    White   Female             0          1721   
3       Asian-Pac-Islander   Female             0             0   
4                    White   Female             0             0   
5                    White     Male             0             0   
...                    ...      ...           ...           ...   
29301                Other     Male             0             0   
29302                Black     Male             0             0   
29303                White     Male             0             0   
29304                White     Male             0             0   
29305                White   Female             0             0   

       hours-per-week  native-country  income  
no                                             
1                  40   United-States       0  
2                  55   United-States       0  
3                  50           Haiti       0  
4                  40   United-States       0  
5                  70   United-States       1  
...               ...             ...     ...  
29301              40     Puerto-Rico       0  
29302              40   United-States       0  
29303              40   United-States       0  
29304              65   United-States       0  
29305              40   United-States       0  

[29305 rows x 15 columns]

In [25]:
test

age          workclass  fnlwgt      education  education-num  \
no                                                                    
29306   18                  ?  245274   Some-college             10   
29307   29            Private   83003        HS-grad              9   
29308   45            Private   35136      Bachelors             13   
29309   42   Self-emp-not-inc   64631      Bachelors             13   
29310   41            Private  195821      Doctorate             16   
...    ...                ...     ...            ...            ...   
48838   45   Self-emp-not-inc  116789        HS-grad              9   
48839   48            Private  185079        HS-grad              9   
48840   63            Private  117473   Some-college             10   
48841   18            Private  150817           11th              7   
48842   31            Private  341632     Assoc-acdm             12   

            marital-status        occupation    relationship    race      sex  \
no                                                                              
29306        Never-married                 ?       Own-child   White     Male   
29307   Married-civ-spouse     Other-service            Wife   White   Female   
29308   Married-civ-spouse      Tech-support         Husband   Black     Male   
29309   Married-civ-spouse   Exec-managerial         Husband   White     Male   
29310   Married-civ-spouse   Exec-managerial            Wife   White   Female   
...                    ...               ...             ...     ...      ...   
48838   Married-civ-spouse      Craft-repair         Husband   White     Male   
48839        Never-married   Exec-managerial   Not-in-family   White   Female   
48840   Married-civ-spouse    Prof-specialty         Husband   White     Male   
48841        Never-married             Sales       Own-child   White   Female   
48842   Married-civ-spouse      Tech-support         Husband   Black     Male   

       capital-gain  capital-loss  hours-per-week  native-country  
no                                                                 
29306             0             0              16   United-States  
29307             0             0              40   United-States  
29308             0             0              40   United-States  
29309             0             0              40   United-States  
29310             0          1902              40   United-States  
...             ...           ...             ...             ...  
48838             0             0              60   United-States  
48839             0             0              50   United-States  
48840          4386             0              40   United-States  
48841             0             0              20   United-States  
48842             0             0              46   United-States  

[19537 rows x 14 columns]

### 1. 전체 데이터 중 결측치 있는 항목(행)을 삭제합니다.
 - 결측치 있는 컬럼 : workclass, occupation, native-country

In [26]:
# rem_MV(remove missing value)에 workclass가 " ?"인 모든 행을 저장합니다.
# drop함수 이용해서 해당 행 삭제합니다.
rem_MV = train[train['workclass'] == " ?"].index
train = train.drop(rem_MV)

In [27]:
rem_MVt = test[test['workclass'] == " ?"].index
test = test.drop(rem_MVt)

In [28]:
# 같은 방식으로 occupation이 "?"인 모든 행을 삭제합니다
rem_MV1 = train[train['occupation'] == " ?"].index
train = train.drop(rem_MV1)

In [29]:
rem_MVt1 = test[test['occupation'] == " ?"].index
test = test.drop(rem_MVt1)

In [30]:
rem_MV2 = train[train['native-country'] == " ?"].index
train = train.drop(rem_MV2)

In [31]:
rem_MVt2 = test[test['native-country'] == " ?"].index
test = test.drop(rem_MVt2)

In [32]:
train

age          workclass  fnlwgt      education  education-num  \
no                                                                    
1       25            Private  219199           11th              7   
2       39            Private   52978   Some-college             10   
3       35            Private  196899      Bachelors             13   
4       64            Private  135527      Assoc-voc             11   
5       24            Private   60783   Some-college             10   
...    ...                ...     ...            ...            ...   
29300   23            Private  117789      Bachelors             13   
29301   20            Private  100605        HS-grad              9   
29302   21            Private  372636        HS-grad              9   
29303   18   Self-emp-not-inc  258474           10th              6   
29304   33            Private  157446           11th              7   

            marital-status          occupation    relationship  \
no                                                               
1                 Divorced   Machine-op-inspct   Not-in-family   
2                 Divorced       Other-service   Not-in-family   
3            Never-married   Handlers-cleaners   Not-in-family   
4                 Divorced        Tech-support   Not-in-family   
5       Married-civ-spouse    Transport-moving         Husband   
...                    ...                 ...             ...   
29300        Never-married     Exec-managerial       Own-child   
29301        Never-married               Sales       Own-child   
29302        Never-married               Sales       Own-child   
29303        Never-married     Farming-fishing       Own-child   
29304        Never-married        Craft-repair   Not-in-family   

                      race      sex  capital-gain  capital-loss  \
no                                                                
1                    White     Male             0             0   
2                    White   Female             0          1721   
3       Asian-Pac-Islander   Female             0             0   
4                    White   Female             0             0   
5                    White     Male             0             0   
...                    ...      ...           ...           ...   
29300                White   Female             0             0   
29301                Other     Male             0             0   
29302                Black     Male             0             0   
29303                White     Male             0             0   
29304                White     Male             0             0   

       hours-per-week  native-country  income  
no                                             
1                  40   United-States       0  
2                  55   United-States       0  
3                  50           Haiti       0  
4                  40   United-States       0  
5                  70   United-States       1  
...               ...             ...     ...  
29300              40   United-States       0  
29301              40     Puerto-Rico       0  
29302              40   United-States       0  
29303              40   United-States       0  
29304              65   United-States       0  

[27173 rows x 15 columns]

In [33]:
test

age          workclass  fnlwgt      education  education-num  \
no                                                                    
29307   29            Private   83003        HS-grad              9   
29308   45            Private   35136      Bachelors             13   
29309   42   Self-emp-not-inc   64631      Bachelors             13   
29310   41            Private  195821      Doctorate             16   
29311   17            Private   25690           10th              6   
...    ...                ...     ...            ...            ...   
48838   45   Self-emp-not-inc  116789        HS-grad              9   
48839   48            Private  185079        HS-grad              9   
48840   63            Private  117473   Some-college             10   
48841   18            Private  150817           11th              7   
48842   31            Private  341632     Assoc-acdm             12   

            marital-status        occupation    relationship    race      sex  \
no                                                                              
29307   Married-civ-spouse     Other-service            Wife   White   Female   
29308   Married-civ-spouse      Tech-support         Husband   Black     Male   
29309   Married-civ-spouse   Exec-managerial         Husband   White     Male   
29310   Married-civ-spouse   Exec-managerial            Wife   White   Female   
29311        Never-married     Other-service       Own-child   White   Female   
...                    ...               ...             ...     ...      ...   
48838   Married-civ-spouse      Craft-repair         Husband   White     Male   
48839        Never-married   Exec-managerial   Not-in-family   White   Female   
48840   Married-civ-spouse    Prof-specialty         Husband   White     Male   
48841        Never-married             Sales       Own-child   White   Female   
48842   Married-civ-spouse      Tech-support         Husband   Black     Male   

       capital-gain  capital-loss  hours-per-week  native-country  
no                                                                 
29307             0             0              40   United-States  
29308             0             0              40   United-States  
29309             0             0              40   United-States  
29310             0          1902              40   United-States  
29311             0             0              10   United-States  
...             ...           ...             ...             ...  
48838             0             0              60   United-States  
48839             0             0              50   United-States  
48840          4386             0              40   United-States  
48841             0             0              20   United-States  
48842             0             0              46   United-States  

[18049 rows x 14 columns]

### 총 29305개의 데이터 중 2132개 데이터 삭제(train)
### 총 19537개의 데이터 중 1488개 데이터 삭제(test)

In [35]:
# train 데이터에서 정답 컬럼
y_train = train['income']

# train 데이터에서 정답 컬럼을 삭제한 후 남은 컬럼들
X_train = train.drop('income', axis = 1)

# 맨 처음에 받았던 test.csv
X_test = test

### 원핫인코딩 - train, test 모두

In [38]:
X_train = pd.get_dummies(X_train)
X_test = pd.get_dummies(X_test)
# X_train
# X_test

age  fnlwgt  education-num  capital-gain  capital-loss  hours-per-week  \
no                                                                              
1       25  219199              7             0             0              40   
2       39   52978             10             0          1721              55   
3       35  196899             13             0             0              50   
4       64  135527             11             0             0              40   
5       24   60783             10             0             0              70   
...    ...     ...            ...           ...           ...             ...   
29300   23  117789             13             0             0              40   
29301   20  100605              9             0             0              40   
29302   21  372636              9             0             0              40   
29303   18  258474              6             0             0              40   
29304   33  157446              7             0             0              65   

       workclass_ Federal-gov  workclass_ Local-gov  workclass_ Private  \
no                                                                        
1                           0                     0                   1   
2                           0                     0                   1   
3                           0                     0                   1   
4                           0                     0                   1   
5                           0                     0                   1   
...                       ...                   ...                 ...   
29300                       0                     0                   1   
29301                       0                     0                   1   
29302                       0                     0                   1   
29303                       0                     0                   0   
29304                       0                     0                   1   

       workclass_ Self-emp-inc  ...  native-country_ Portugal  \
no                              ...                             
1                            0  ...                         0   
2                            0  ...                         0   
3                            0  ...                         0   
4                            0  ...                         0   
5                            0  ...                         0   
...                        ...  ...                       ...   
29300                        0  ...                         0   
29301                        0  ...                         0   
29302                        0  ...                         0   
29303                        0  ...                         0   
29304                        0  ...                         0   

       native-country_ Puerto-Rico  native-country_ Scotland  \
no                                                             
1                                0                         0   
2                                0                         0   
3                                0                         0   
4                                0                         0   
5                                0                         0   
...                            ...                       ...   
29300                            0                         0   
29301                            1                         0   
29302                            0                         0   
29303                            0                         0   
29304                            0                         0   

       native-country_ South  native-country_ Taiwan  \
no                                                     
1                          0                       0   
2                          0                       0   
3                          0                       0   
4                 

In [39]:
!pip install lightgbm

In [40]:
!pip install xgboost

In [49]:
from sklearn.model_selection import cross_val_score, train_test_split, KFold
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier

# 모델 설정
# ------------------------

lgb_model = LGBMClassifier(n_estimators = 1000, learning_rate=0.01, random_state=123)
lgb_model.fit(X_train, y_train) # transform_X_train

xg_model = XGBClassifier(n_estimators = 100, learning_rate=0.01, random_state = 123)
xg_model.fit(X_train, y_train) # transform_X_train

forest_model = RandomForestClassifier(n_estimators=100, random_state= 11)
forest_model.fit(X_train, y_train) # transform_X_train

# ---------------------------------------------

# 정확도

kfold = KFold(n_splits=10, random_state=123, shuffle=True)

acc_result = cross_val_score(lgb_model, # 모델
                         X_train, # transform_X_train
                         y_train,
                         cv = kfold,
                         scoring='f1'
                        )

print(acc_result.mean())

# -----------------------------------------------

# F1 score

f1_result = cross_val_score(lgb_model, # 모델
                         X_train, # transform_X_train
                         y_train,
                         cv = kfold,
                         scoring='f1'
                        )

print(f1_result.mean())

# ------------------------------------------

C:\Users\qoska\anaconda3\lib\site-packages\xgboost\sklearn.py:892: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[15:55:53] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
0.7166345576250718
0.7166345576250718


In [45]:
# 정확도

kfold = KFold(n_splits=10, random_state=123, shuffle=True)

acc_result = cross_val_score(xg_model, # 모델
                         X_train, # transform_X_train
                         y_train,
                         cv = kfold,
                         scoring='f1'
                        )

print(acc_result.mean())

# -----------------------------------------------

# F1 score

f1_result = cross_val_score(xg_model, # 모델
                         X_train, # transform_X_train
                         y_train,
                         cv = kfold,
                         scoring='f1'
                        )

print(f1_result.mean())

# ------------------------------------------

[15:49:41] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:49:54] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:50:04] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:50:16] WARNING: C:/Users/Administrator/workspace/xgboost-win64_rel

In [50]:
# 정확도

kfold = KFold(n_splits=10, random_state=123, shuffle=True)

acc_result = cross_val_score(forest_model, # 모델
                         X_train, # transform_X_train
                         y_train,
                         cv = kfold,
                         scoring='f1'
                        )

print(acc_result.mean())

# -----------------------------------------------

# F1 score

f1_result = cross_val_score(forest_model, # 모델
                         X_train, # transform_X_train
                         y_train,
                         cv = kfold,
                         scoring='f1'
                        )

print(f1_result.mean())

# ------------------------------------------

0.67864532211554
0.67864532211554
